src/SSTClosGovernor.cpp

```diff
@@ -0,0 +1,196 @@
+        // turbo range, but give high priority to distribute up to max.
+        m_platform_io.write_control("SST::COREPRIORITY:0:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 0.0);
+        m_platform_io.write_control("SST::COREPRIORITY:0:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:0:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_max);
+
+        // Next-highest bucket. Apply the same ranges, but with less priority.
+        m_platform_io.write_control("SST::COREPRIORITY:1:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 0.34);
+        m_platform_io.write_control("SST::COREPRIORITY:1:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:1:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_max);
+
+        // First low-priority bucket. Initially just give it the bottom of the
+        // turbo range. Potentially go lower at run time.
+        m_platform_io.write_control("SST::COREPRIORITY:2:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 0.67);
+        m_platform_io.write_control("SST::COREPRIORITY:2:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:2:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, (m_frequency_sticker + m_frequency_max) / 2);
+
+        // Least prioritized bucket. Initially just give it the bottom of the
+        // turbo range. Potentially go lower at run time.
+        m_platform_io.write_control("SST::COREPRIORITY:3:PRIORITY",
+                                    GEOPM_DOMAIN_BOARD, 0, 1.0);
+        m_platform_io.write_control("SST::COREPRIORITY:3:FREQUENCY_MIN",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+        m_platform_io.write_control("SST::COREPRIORITY:3:FREQUENCY_MAX",
+                                    GEOPM_DOMAIN_BOARD, 0, m_frequency_sticker);
+    }
+
+    int SSTClosGovernorImp::clos_domain_type(void) const
+    {
+        return m_clos_assoc_ctl_domain_type;
+    }
+
+    void SSTClosGovernorImp::adjust_platform(
+        const std::vector<double> &clos_by_core)
+    {
+        if (clos_by_core.size() != m_clos_control_idx.size()) {
+            throw Exception("SSTClosGovernorImp::" + std::string(__func__) +
+                                "(): size of request vector does not match size of control domain.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        m_do_write_batch = m_is_enabled && clos_by_core != m_last_clos;
+        if (m_is_enabled) {
+            for (size_t idx = 0; idx < m_clos_control_idx.size(); ++idx) {
```

Comment:
What if you move the `size()` call out of the loop, to avoid it getting called during every single iteration? You have one call instead of **N** of them.
```cpp
szie_t m_clos_control_idx_size = m_clos_control_idx.size();
for (size_t idx = 0; idx < m_clos_control_idx_size; ++idx) {}
```